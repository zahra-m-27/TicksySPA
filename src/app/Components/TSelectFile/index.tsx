import Assets from '../../Assets';
import styles from './styles.module.scss';
import ClassNames from '../../Utilities/ClassNames';
import {useRef} from 'react';

interface Props {
  className?: string;
  onRemoveAttachment?: () => void;
  onSelectFile?: (file: File) => void;
}

export default function TSelectFile({
  className,
  onSelectFile,
  onRemoveAttachment,
}: Props) {
  const inputRef = useRef<HTMLInputElement>(null);

  const onChange = (e: any) => {
    if (e.target.files && onSelectFile) {
      onSelectFile(e.target.files[0]);
    }
  };

  const onRemoveFile = () => {
    if (inputRef.current) {
      inputRef.current.value = '';
      inputRef.current.files = null;
    }
    if (onRemoveAttachment) {
      onRemoveAttachment();
    }
  };

  const container_style = ClassNames(styles.container, className);

  return (
    <div className={container_style}>
      <p className={styles.label}>پيوست فايل</p>
      <div className={styles.select_file}>
        <img
          alt="delete"
          data-testid="remove-file"
          onClick={onRemoveFile}
          src={Assets.SVGs.Delete}
          className={styles.delete}
        />
        <input
          ref={inputRef}
          type="file"
          data-testid="file-input"
          className={styles.file}
          onChange={onChange}
        />
      </div>
    </div>
  );
}
