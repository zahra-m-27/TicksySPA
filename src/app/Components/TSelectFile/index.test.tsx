import React from 'react';
import {render} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import TSelectFile from './index';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import user from '@testing-library/user-event';

const onSelectFile = jest.fn();
const onRemoveAttachment = jest.fn();

const sample = (
  <TSelectFile
    onSelectFile={onSelectFile}
    onRemoveAttachment={onRemoveAttachment}
  />
);

test('render in ReactDOM', () => {
  const div = document.createElement('div');
  ReactDOM.render(sample, div);
});

test('select file', () => {
  const {getByTestId} = render(sample);
  const fileInputElement = getByTestId(/file-input/i) as HTMLInputElement;

  const file = new File(['select'], '', {type: 'file'});
  userEvent.upload(fileInputElement, file);

  expect(fileInputElement.files).toBeTruthy();
  if (fileInputElement.files) {
    expect(fileInputElement.files[0]).toStrictEqual(file);
    expect(fileInputElement.files.item(0)).toStrictEqual(file);
  }
  expect(fileInputElement.files).toHaveLength(1);

  expect(onSelectFile).toBeCalledTimes(1);
  expect(onSelectFile).toBeCalledWith(file);
});

test('remove file', () => {
  const {getByTestId} = render(sample);
  const fileInputElement = getByTestId(/file-input/i) as HTMLInputElement;

  const file = new File(['select'], '', {type: 'file'});
  userEvent.upload(fileInputElement, file);

  expect(fileInputElement.files).toBeTruthy();
  if (fileInputElement.files) {
    expect(fileInputElement.files[0]).toStrictEqual(file);
    expect(fileInputElement.files.item(0)).toStrictEqual(file);
  }
  expect(fileInputElement.files).toHaveLength(1);

  expect(onSelectFile).toBeCalledTimes(1);
  expect(onSelectFile).toBeCalledWith(file);

  const removeFileElement = getByTestId(/remove-file/i);
  user.click(removeFileElement);

  expect(fileInputElement.files).toBeFalsy();
});

it('snapshot', () => {
  const snapshot = renderer.create(sample).toJSON();
  expect(snapshot).toMatchSnapshot();
});
