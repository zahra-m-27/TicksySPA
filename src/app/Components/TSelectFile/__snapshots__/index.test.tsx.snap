// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
<div
  className="container"
>
  <p
    className="label"
  >
    پيوست فايل
  </p>
  <div
    className="select_file"
  >
    <img
      alt="delete"
      className="delete"
      data-testid="remove-file"
      onClick={[Function]}
      src="delete.svg"
    />
    <input
      className="file"
      data-testid="file-input"
      onChange={[Function]}
      type="file"
    />
  </div>
</div>
`;
